WallarooLabs/wallaroo
# Add AtLeastOnceSourceConnector

- includes at_least_once_feed example in examples/python/celsius_connector
- includes mock server in examples/python/celsius_connector
- includes file based celsius data sender (multi-file, multi-stream)

part of #2719

# Testing instructions

When running either the `mock_at_least_once_server` or `at_least_once_feed`, add wallaroo to your pythonpath:
`export PYTHONPATH=~/wallaroo-tutorial/wallaroo/machida/lib` (you might need to change `~` to explicit `/home/user/...`)

1. Review the core code:
    - [ ] `machida/lib/wallaroo/experimental/__init__.py`
    - [ ] `machida/lib/wallaroo/experimental/connector_wire_messages.py`
2. Look at the example connector in `examples/python/celsius_connector`
    - [ ] `at_least_once_feed` (see run instructions in its docstring)
    - [ ] `mock_at_least_once_server`
3. Confirm that the protocol behaviour is correct, where relevant:
    - [ ] hello / ok / error
    - [ ] notify / notify_ack / stream_updated / reset
    - [ ] eos
4. Look at the way the actual connecting and sending is implemented in the example connector:
    - Does the use of `self.joining()` make sense as an initial blocking check before the notify_acks arrive?
    - Does the use of the `BaseIter` class make sense (specifically for "consumable" objects like static files, sequences, and generators)?
    - Is there anything you would like to change about this example?
5. Review the ergonomics of the `AtLeastOnceSourceConnector` class in the context of an actual connector:
    - [ ] Does the overall design make sense to you?
    - [ ] Do you love / like / dislike / hate the subclass + user-definable methods approach with `AtLeastOnceSourceConnector`?
    - Do you think any of these user-defined methods could be omitted?
        - `stream_updated(stream)`
        - `handle_invalid_message(msg)`
        - `handle_restart()`
        - `handle_error()`
    - Are there any user-defined methods you think we should add?
6. Any other thoughts or feedback?
